### PR TITLE
Filter lower versions when checking ignored versions

### PR DESCRIPTION
**Pull request by feelepxyz**
Changes the condition that raises `AllVersionsIgnored` when all new
versions are ignored when there are some new versions, filtering out
lower versions.

I opted to do this in the condition to not introduce any breaking
changes in the version checkers.

I've also explored always filtering lower version before filtering
ignored versions but this has been implemented inconsistently across
package managers so could cause a bunch of subtle bugs.

Most package managers only filter lower versions when checking for the
lowest security fix version, and rely on the latest version being set,
even if it's the current version. 🤦‍♂️

I wonder if it would make more sense to do a separate refactor to start
using the `UpdateConfig` and `IgnoreCondition` classes in a more generic
way in the base update checker instead of doing the same filtering in each
package manager.

## TODO:
- [x] Add some test cases

### CODE REVIEW

File: hex/lib/dependabot/hex/update_checker.rb [hunk 1]
```diff
@@ -221,13 +221,23 @@ def latest_release_from_hex_registry
...
+            end
 
             filtered.max
           end
       end
       # rubocop:enable Metrics/PerceivedComplexity
 
+      def filter_lower_versions(versions_array)
+        versions_array.select do |version|
+          next unless current_version
```
Comment (thepwagner) [inline hex/lib/dependabot/hex/update_checker.rb:235]:
```suggestion
        return [] unless current_version

        versions_array.select do |version|
```

Nit: `current_version` isn't going to change inside the loop, so check it once and exit early.

Comment (thepwagner) [reply hex/lib/dependabot/hex/update_checker.rb:235]: Nit: should we care about the memoization of `current_version`? We could shave some nanos by storing it outside the loop.

File: hex/lib/dependabot/hex/update_checker.rb [hunk 2]
```diff
@@ -246,13 +256,14 @@ def hex_registry_response
...
+      end
+
       def wants_prerelease?
-        current_version = dependency.version
-        if current_version &&
-           version_class.correct?(current_version) &&
-           version_class.new(current_version).prerelease?
-          return true
-        end
+        return true if current_version&.prerelease?
```
Comment (thepwagner) [inline hex/lib/dependabot/hex/update_checker.rb:266]: Noice: this cleanup!

**Review by thepwagner — APPROVED** [review]
Diff LGTM.

Can you walk me through a case where this is new behaviour is desirable?
Maybe we're ignoring `>= 0`, but are currently on the latest version.

**Comment by feelepxyz** [comment]
@thepwagner I've attempted to fix this issue where currently we won't raise `AllVersionsIgnored ` when ignoring update types as these only filter new versions:

Pre:
Screenshot 2021-05-06 at 14 25 51

Post:
Screenshot 2021-05-06 at 14 25 03

So now, if the ignore conditions end up rejecting all new versions we raise.

File: common/spec/dependabot/git_commit_checker_spec.rb [hunk 3]
```diff
@@ -944,6 +963,18 @@
           end
         end
 
+        context "all later versions ignored" do
```
Comment (mctofu) [inline common/spec/dependabot/git_commit_checker_spec.rb:966]: This test is currently not passing because `current_version` ends up being nil

**Comment by mctofu** [comment]
> [ ] Add some test cases

I made it through about half of the ecosystems. The `GitCommitChecker` is not yet passing but I haven't dug in yet to see if my test setup is wrong or if target code needs some adjustment.

**Review by jasonrudolph — APPROVED** [review]
> I added three test cases to each of the ecosystems to check that when raise_on_ignored is enabled...

I reviewed all the tests and they look good.

Thanks so much for taking the time to [commit the tests separately for each ecosystem](https://github.com/dependabot/dependabot-core/compare/f034763fa7938ffe039768fd3390e2175265d19a...db79fbe667bd53c3cf09e30b5430cb2f2b4fff82). For me, that made the review nice and approachable, as I was able to review small, focused commits instead of facing the whole diff at once. 😅

**Pull request merged**